klippy/extras/exclude_object.py

```diff
@@ -0,0 +1,263 @@
+            desc=self.cmd_END_CURRENT_OBJECT_help)
+        self.gcode.register_command(
+            'EXCLUDE_OBJECT', self.cmd_EXCLUDE_OBJECT,
+            desc=self.cmd_EXCLUDE_OBJECT_help)
+        self.gcode.register_command(
+            'EXCLUDE_OBJECT_RESET', self.cmd_EXCLUDE_OBJECT_RESET,
+            desc=self.cmd_EXCLUDE_OBJECT_RESET_help)
+        self.gcode.register_command(
+            'DEFINE_OBJECT', self.cmd_DEFINE_OBJECT,
+            desc=self.cmd_DEFINE_OBJECT_help)
+        self.gcode.register_command(
+            'LIST_OBJECTS', self.cmd_LIST_OBJECTS,
+            desc=self.cmd_LIST_OBJECTS_help)
+        self.gcode.register_command(
+            'LIST_EXCLUDED_OBJECTS', self.cmd_LIST_EXCLUDED_OBJECTS,
+            desc=self.cmd_LIST_EXCLUDED_OBJECTS_help)
+    def _setup_transform(self):
+        if not self.next_transform:
+            tuning_tower = self.printer.lookup_object('tuning_tower')
+            if tuning_tower.is_testing():
+                logging.info('The ExcludeObject move transform is not being '
+                    'loaded due to Tuning tower being Active')
+                return
+
+            self.next_transform = self.gcode_move.set_move_transform(self,
+                                                                     force=True)
+            self.last_position[:] = self.get_position()
+            self.last_position_extruded[:] = self.get_position()
+            self.last_position_excluded[:] = self.get_position()
+            self.extrusion_offsets = [[0., 0., 0., 0.]]
+            self.max_position_extruded = 0
+            self.max_position_excluded = 0
+            self.extruder_adj = 0
+            self.objects = {}
+            self.excluded_objects = []
+            self.current_object = None
+
+    def _handle_connect(self):
+        self.toolhead = self.printer.lookup_object('toolhead')
+
+    def _reset_file(self):
+        self.objects = {}
+        self.excluded_objects = []
+        self.current_object = None
+        if self.next_transform:
+            tuning_tower = self.printer.lookup_object('tuning_tower')
+            if tuning_tower.is_testing():
+                logging.info('The Exclude Object move transform was not '
+                    'unregistered because it is not at the head of the '
+                    'transform chain.')
+                return
+
+            self.gcode_move.set_move_transform(self.next_transform, force=True)
+            self.next_transform = None
+            self.gcode_move.reset_last_position()
+
+    def get_position(self):
+        extr_idx = self.toolhead.get_extruder().get_name()[8:]
+        if extr_idx:
+            self.extruder_idx = int(extr_idx)
+        else:
+            self.extruder_idx = 0
+        len_diff = self.extruder_idx + 1 - len(self.extrusion_offsets)
+        if len_diff > 0:
+            self.extrusion_offsets.extend([[0., 0., 0., 0.]] * len_diff)
+        offset = self.extrusion_offsets[self.extruder_idx]
```

Comment:
I'm not sure what this code does, but it looks a bit unusual to me.  I suspect things would be more clear (and more efficient) if `self.extrusion_offsets` was a dict instead of a list - then the code could just be `offset = self.extrusion_offsets[self.toolhead.get_extruder().get_name()]`.

Reply:
This was my suggestion, the goal is to track the amount of extrusion we are excluding per extruder.  I agree that your suggestion is a cleaner way of handling it.